Improve score_fusion_memory()

aorenste opened the pull request:
Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

`score_fusion_memory` is called O(n^2) times - so by moving the set union, `has_unbacked_symbols` check, and `numbytes_hint` out of the loop we call them O(n) times and the O(n^2) call gets cheaper.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s
After this change: 10m15s

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127061
* __->__ #127060


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

aorenste committed:
Improve score_fusion_memory()

Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s real
After this change: 10m15s real

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -246,9 +250,15 @@ def has_aliasing_or_mutation(self) -> bool:
         self.prune_deps()
+        # common_deps are a summary of read_writes used by score_fusion_memory()
+        self.common_deps = {
+            (dep, hint)
```
NIT: This would make more sense as a dictionary `dep_to_numbytes_hint`. As lookup by hint doesn't make sense and at this stage there is nothing "common" as it only refers to a single node.

aorenste replied:
"common" refers to the fact that it's both read & write (maybe `all_deps` or `score_fusion_memory_deps` would be a better name?)

Using a dictionary would be slower because we need to find the intersection between the deps of two nodes - so if it were a dict we'd have to construct the `keys()` set at the point of comparison.

peterbell10 replied:
> "common" refers to the fact that it's both read & write (maybe all_deps or score_fusion_memory_deps would be a better name?)

No, common refers to it being the dependencies that the two nodes have in common with each other. The new name is fine though.

> Using a dictionary would be slower because we need to find the intersection between the deps of two nodes - so if it were a dict we'd have to construct the keys() set at the point of comparison.

Hrm okay, I didn't know that dict keys objects didn't support union natively.

aorenste replied:
I imagine that they do - but you then have to construct keys objects in O(n^2) calls to `score_fusion_memory_deps`.

aorenste committed:
minor update + rebase on "Improve score_fusion_memory()"

Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

`score_fusion_memory` is called O(n^2) times - so by moving the set union, `has_unbacked_symbols` check, and `numbytes_hint` out of the loop we call them O(n) times and the O(n^2) call gets cheaper.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s
After this change: 10m15s



[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -159,6 +159,10 @@ def should_prune(dep: Dep) -> bool:
+    read_writes: dependencies.ReadWrites
+    unmet_dependencies: Set[Dep]
+    # Processed deps used while scoring fusion
+    common_deps: Set[Tuple[Dep, int]]
```
NIT: This is stale now.

aorenste committed:
Update on "Improve score_fusion_memory()"

Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

`score_fusion_memory` is called O(n^2) times - so by moving the set union, `has_unbacked_symbols` check, and `numbytes_hint` out of the loop we call them O(n) times and the O(n^2) call gets cheaper.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s
After this change: 10m15s



[ghstack-poisoned]

aorenste committed:
Update on "Improve score_fusion_memory()"

Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

`score_fusion_memory` is called O(n^2) times - so by moving the set union, `has_unbacked_symbols` check, and `numbytes_hint` out of the loop we call them O(n) times and the O(n^2) call gets cheaper.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s
After this change: 10m15s



[ghstack-poisoned]

aorenste committed:
Update on "Improve score_fusion_memory()"

Related to #98467

The tacotron2 benchmark creates a lot of nodes which fusion then checks. This
improves some of the perf of that checking.

`score_fusion_memory` is called O(n^2) times - so by moving the set union, `has_unbacked_symbols` check, and `numbytes_hint` out of the loop we call them O(n) times and the O(n^2) call gets cheaper.

Testing:
```
time python benchmarks/dynamo/torchbench.py --accuracy --inference --amp --backend inductor --disable-cudagraphs --device cuda --only tacotron2
```

Before this change: 12m33s
After this change: 10m15s



[ghstack-poisoned]
